lib/services/role.go

```diff
@@ -2092,6 +2120,59 @@ func (set RoleSet) EnhancedRecordingSet() map[string]bool {
 	return m
 }
 
+// HostUsers returns host user information matching a server or nil if
+// a role disallows host user creation
+func (set RoleSet) HostUsers(s types.Server) (*HostUsersInfo, error) {
+	groups := map[string]struct{}{}
+	sudoers := map[string]struct{}{}
+	serverLabels := s.GetAllLabels()
+	for _, role := range set {
+		result, _, err := MatchLabels(role.GetNodeLabels(types.Allow), serverLabels)
+		if err != nil {
+			return nil, err
+		}
+		// skip nodes that dont have matching labels
+		if !result {
+			continue
+		}
+		createHostUser := role.GetOptions().CreateHostUser
+		// if any of the matching roles do not enable create host
+		// user, the user should not be allowed on
+		if createHostUser == nil || !createHostUser.Value {
+			return nil, trace.AccessDenied("user is not allowed to create host users")
+		}
+		for _, group := range role.GetHostGroups(types.Allow) {
+			groups[group] = struct{}{}
+		}
+		for _, group := range role.GetHostGroups(types.Allow) {
+			groups[group] = struct{}{}
+		}
+		for _, sudoer := range role.GetHostSudoers(types.Allow) {
+			sudoers[sudoer] = struct{}{}
+		}
+	}
+	for _, role := range set {
+		result, _, err := MatchLabels(role.GetNodeLabels(types.Deny), serverLabels)
+		if err != nil {
+			return nil, err
+		}
+		if !result {
+			continue
+		}
+		for _, group := range role.GetHostGroups(types.Deny) {
+			delete(groups, group)
+		}
+		for _, sudoer := range role.GetHostSudoers(types.Deny) {
+			delete(sudoers, sudoer)
+		}
```

Comment:
I'm not sure it makes sense to be able to deny a specific sudoers entry that must match exactly, it seems error-prone. Maybe we should specify that you can only pass a wildcard like the following and that will deny all sudo entries? Just a thought, I'm not sure even that makes sense
```yaml
kind: role
version: v5
metadata:
  name: no-sudo
spec:
  deny:
    host_sudoers: ['*']
```

Reply:
I think this is the only outstanding item here. Having this line here for now is probably fine (because I doubt anyone will use it) and I don't want to block the PR but it doesn't make much sense. @r0mant do we need a good way to deny `host_sudoers` in the role? Use an option?